core/security_group_rule_test.go

```diff
@@ -0,0 +1,802 @@
+				wantPath: "/core/v1/security_groups/rules/_",
+				wantValues: url.Values{
+					"security_group_rule[id]": []string{"123"},
+				},
+				wantMethod: "GET",
+				wantBody:   nil,
+				doErr:      fmt.Errorf("flux capacitor undercharged"),
+				doResp:     &katapult.Response{},
+			},
+			wantErr: "flux capacitor undercharged",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := NewSecurityGroupRulesClient(&fakeRequestMaker{
+				t:    t,
+				args: tt.frm,
+			})
+
+			got, resp, err := c.Get(
+				context.Background(),
+				tt.args.ref,
+			)
+			assert.Equal(t, tt.frm.doResp, resp)
+
+			if tt.wantErr == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.EqualError(t, err, tt.wantErr)
+			}
+
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestSecurityGroupRulesClient_GetByID(t *testing.T) {
+	type args struct {
+		id string
+	}
+	tests := []struct {
+		name    string
+		frm     fakeRequestMakerArgs
+		args    args
+		want    *SecurityGroupRule
+		wantErr string
+	}{
+		{
+			name: "success",
+			args: args{
+				id: "123",
+			},
+			want: &SecurityGroupRule{
+				ID:        "123",
+				Direction: "inbound",
+			},
+			frm: fakeRequestMakerArgs{
+				wantPath: "/core/v1/security_groups/rules/_",
+				wantValues: url.Values{
+					"security_group_rule[id]": []string{"123"},
+				},
+				wantMethod: "GET",
+				wantBody:   nil,
+				doResponseBody: &SecurityGroupRulesResponseBody{
+					SecurityGroupRule: &SecurityGroupRule{
+						ID:        "123",
+						Direction: "inbound",
+					},
+				},
+				doResp: &katapult.Response{},
+			},
+		},
+		{
+			name: "new request fails",
+			args: args{
+				id: "123",
+			},
+			frm: fakeRequestMakerArgs{
+				wantPath: "/core/v1/security_groups/rules/_",
+				wantValues: url.Values{
+					"security_group_rule[id]": []string{"123"},
+				},
+				wantMethod: "GET",
+				wantBody:   nil,
+				newReqErr:  fmt.Errorf("rats chewed cables"),
+			},
+			wantErr: "rats chewed cables",
+		},
+		{
+			name: "http do fails",
+			args: args{
+				id: "123",
+			},
+			frm: fakeRequestMakerArgs{
+				wantPath: "/core/v1/security_groups/rules/_",
+				wantValues: url.Values{
+					"security_group_rule[id]": []string{"123"},
+				},
+				wantMethod: "GET",
+				wantBody:   nil,
+				doErr:      fmt.Errorf("flux capacitor undercharged"),
+				doResp:     &katapult.Response{},
+			},
+			wantErr: "flux capacitor undercharged",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := NewSecurityGroupRulesClient(&fakeRequestMaker{
+				t:    t,
+				args: tt.frm,
+			})
+
+			got, resp, err := c.GetByID(
+				context.Background(),
+				tt.args.id,
+			)
+			assert.Equal(t, tt.frm.doResp, resp)
+
+			if tt.wantErr == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.EqualError(t, err, tt.wantErr)
+			}
+
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestSecurityGroupRulesClient_List(t *testing.T) {
```

Comment:
Very minor nitpick, but I like ordering the tests for different functions in the same order as the function themselves in the implementation file :)